src/HTTPserver.cpp

```diff
@@ -35,6 +35,101 @@ static HTTPserver *httpserver;
 
 /* ****************************************** */
 
+/* The total number of threads to use in mongoose */
+#define SERVER_NUM_THREADS 16
+
+/* The maximum number of threads to dedicate do WebSocket connections */
+#define SERVER_MAX_WEBSOCKET_THREADS 4
+
+/* The maximum number of active connections (busy threads) per host */
+#define SERVER_MAX_THREADS_PER_HOST 8
+
+/* ****************************************** */
+
+class ConnectionsLimiter {
+  private:
+    Mutex *mutex;
+    int active_threads;
+    int active_websockets;
+    struct {
+      uint32_t ip;
+      int port;
+      bool websocket;
+    } active_hosts[SERVER_NUM_THREADS];
+
+  public:
+    ConnectionsLimiter() {
+      mutex = new Mutex();
+      active_threads = 0;
+      active_websockets = 0;
+      memset(active_hosts, 0, sizeof(active_hosts));
+    }
+
+    ~ConnectionsLimiter() {
+      delete mutex;
+    }
+
+    bool connectHost(uint32_t ip, int port, bool is_websocket) {
+      bool success = true;
+
+      mutex->lock(__FILE__, __LINE__);
+
+      if (((is_websocket) && (active_websockets >= SERVER_MAX_WEBSOCKET_THREADS)) ||
+          (active_threads >= SERVER_NUM_THREADS)) {
+        success = false;
+      } else {
+        int num_hits = 0;
+        for (int i=0; i<active_threads; i++) {
+          if (active_hosts[i].ip == ip)
+            num_hits++;
+        }
+
+        if (num_hits >= SERVER_MAX_THREADS_PER_HOST)
+          success = false;
+      }
+
+      if (success) {
+        active_hosts[active_threads].ip = ip;
+        active_hosts[active_threads].port = port;
+
+        if (is_websocket) {
+          active_hosts[active_threads].websocket = true;
+          active_websockets++;
+        } else {
+          active_hosts[active_threads].websocket = false;
+        }
+        
+        active_threads++;
+      }
+
+      mutex->unlock(__FILE__, __LINE__);
+
+      return success;
+    }
+
+    void disconnectHost(uint32_t ip, int port) {
+      mutex->lock(__FILE__, __LINE__);
+
+      for (int i=0; i<active_threads; i++) {
+        if ((active_hosts[i].ip == ip) && (active_hosts[i].port == port)) {
+          bool is_websocket = active_hosts[i].websocket;
+
+          memmove(&active_hosts[i], &active_hosts[i+1], (active_threads - i - 1) * sizeof(active_hosts[0]));
```

Comment:
ISO/IEC 9899:1999:

$7.20.5
>  These utilities make use of a comparison function to search or sort arrays of unspecified type. Where an argument declared as size_t nmemb specifies the length of the array for a function, nmemb can have the value zero on a call to that function; [...] Pointer arguments on such a call shall still have valid values, as described in 7.1.4.

7.1.4
> If an argument to a function has an invalid value (such as a value outside the domain of the function, or a pointer outside the address space of the program, or a null pointer, or a pointer to non-modifiable storage when the corresponding parameter is not const-qualified) or a type (after promotion) not expected by a function with variable number of arguments, the behavior is undefined.

So even if a number of bytes equal to zero is an acceptable value, the out-of-bounds pointer is not.